accepted/2021/runtime-security-mitigations.md

```diff
@@ -0,0 +1,213 @@
+
+IL code interpreters circumvent the environment restriction on runtime code generation and allow .NET libraries that depend on runtime code generation to work, with lower performance. From a security point of view, IL code interpreters are not desirable in environments with disallowed code generation since IL code is equivalent of executable code as described above.
+
+## Out-of-process Jit-In-Time Compiler (JIT)
+
+Moving code generation into separate process has been implemented as a security mitigation in some cases, most recently in Chakra JavaScript engine. With this mitigation, the regular process does not have privilege to generate new code and it must delegate the code generation to dedicated process instead. The dedicated process with special privileges to introduce dynamically generated code is audited and heavily scrutinized. The advantage of this solution is that no code in the original process ever needs to write executable memory, so entire attack vectors (like compromised code altering memory mappings) can be shut down entirely. [Google Project Zero analyzed the Chakra implementation](https://github.com/googleprojectzero/p0tools/blob/master/JITServer/JIT-Server-whitepaper.pdf) and pointed out some of its weaknesses.
+
+We do not see strong need for introducing out of process JIT for .NET runtime in foreseeable future. The primary reason is that .NET does not support sand-boxed execution of untrusted code anymore. The out-of-proc JIT has advantages for execution of untrusted (verifiable) code. The security benefits of out-of-proc JIT for trusted unverifiable IL code are less interesting since strong soundness validation is not feasible for unverifiable IL code. The secondary reasons include availability of more secure options that disallow runtime code generation altogether, the implementation cost and necessity of a second process to run alongside the primary process.
+
+# Control-flow Integrity
+
+Once attackers are no longer able to inject arbitrary code into the process thanks to executable space protections, they start looking for creative ways to leverage existing code in the process to mount the attack. [Return-oriented Programming](https://en.wikipedia.org/wiki/Return-oriented_programming) (ROP) and [Jump-oriented Programming](https://www.csc2.ncsu.edu/faculty/xjiang4/pubs/ASIACCS11.pdf) (JOP) are generalized exploit techniques that take advantage of situations where the program uses indirection for control transfer. The affected indirection is return address for ROP and function pointer for JOP. Control-flow integrity adds validation of these indirections to eliminate or reduce chance of them being used to mount successful attacks.
+
+The control-flow integrity validations are emerging area. They are not yet standardized across operating systems or hardware vendors.
```

Comment:
NIT: "an" emerging area.

Reply:
Are you saying here we won't support them unless they're standardized? That's what the last sentence insinuates.

Reply:
I believe this refers more to the concept that it isn't feasible to achieve mitigations with the matching security characteristics across operating systems and hardware vendors as the capabilities are not standardized enough. In particular, strong mitigations in this space typically require involvement with the OS in ways that are fundamentally different for each OS, and the hardware mitigations in the space are only available on a subset of hardware and also have tight OS integration that differs from OS to OS making finding a least common denominator api concept is difficult.